### PR TITLE
proposal: Macvtap mode for pod networking

### PR DESCRIPTION
**Pull request by kvaps**
- Rendered version: [macvtap-mode-for-pod-networking.md](https://github.com/kvaps/community/blob/macvtap-mode-for-pod-networking/design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md)
- Implementation: https://github.com/kubevirt/kubevirt/pull/7648
- Slack thread:
https://kubernetes.slack.com/archives/C8ED7RKFE/p1650903781064009?thread_ts=1650477074.237059&cid=C8ED7RKFE

### CODE REVIEW

File: design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+Also this change extends existing macvtap driver to write cached information about the interface and prepare DHCP config.
+
+**How is the macvtap device created?**
+
+The new flag is added to the **tap-device-maker** binary. If it is called with `--parent-name` specified, then macvtap device will be created instead of tap.  
+But unlike tap, before running VM with macvtap interface, you need need to:
+
+- Set device rule for the cgroup running this pod:
+
+  *more details here: https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt*
```
Comment (maiqueb) [inline design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md:97]: Will you have to write different device rules for cgroups-v1 / v2 ?

File: design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+- Create character device in pod mount namespace:
+  ```
+  cat /sys/class/net/tap0/macvtap/tap5/dev
+  238:1
+  mknod /dev/tap5 c 238 1
+  ```
+
+## API Examples
+Thanks to macvtap-cni we already have everything needed and we don't need to modify the API.  
+From the user perspective we just allow to using macvtap mode to bind pod networking, example:
```
Comment (maiqueb) [inline design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md:108]: maybe add more detail into how you will "allow using macvtap mode to bind pod networking" - i.e. [updating the admitter](https://github.com/kubevirt/kubevirt/blob/392ee0cae3e8477ca1c158984eed44faa8a184a3/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L89).

Could be OK to leave as is - that implementation detail is fairly obvious.

**Review by maiqueb — COMMENTED** [review]
A question + comment.

File: design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md (hunk 1 above)
Comment (kvaps) [reply design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md:97]: No, I just reused existing cgroup manager interface:
https://github.com/kubevirt/kubevirt/blob/1daa0f3219aaa69891383b9c1d0158ae4fe15e91/pkg/network/driver/common.go#L446-L467

It should be agnostic to cgroups version:
https://github.com/kubevirt/kubevirt/blob/3bee69d3e9ce9cf85e97ff969a07eb29cf328d04/pkg/virt-handler/cgroup/cgroup.go#L42-L46

Meanwhile we use the same logic for hot-plugging devices:
https://github.com/kubevirt/kubevirt/blob/417fce4b5643e341a91d2fbbd2c316c2e9ca4ae6/pkg/virt-handler/hotplug-disk/mount.go#L490-L500

File: design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md (hunk 2 above)
Comment (kvaps) [reply design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md:108]: Do you mean to add some option like `permitBridgeInterfaceOnPodNetwork: true`?
I think this is not required. Unlike bridge this method way more safe.

From the node perspective traffic from another side of veth interface no different from the normal Pod without VM.
The MAC address will always be filtered to the single one used in macvtap interface.

Comment (maiqueb) [reply design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md:108]: I simply mean the current admitter code will reject creating a VMI having a macvtap interface meant for pod network - afaiu,  "allow using macvtap mode to bind pod networking" means relaxing that rule, and allow macvtap interface to be created also for the pod network **if** the feature gate is enabled.

**Comment by kvaps** [comment]
I migrated my fork from `kvaps/community` to `deckhouse/3p-kubevirt-community`, now it will be supported by deckhouse project
fyi

**Comment by EdDev** [comment]
It is now possible to introduce a new binding using the network binding plugins infrastructure.
There are examples for slirp, passt (merged) and macvtap (in merge queue at the time of writing this message).

If this is still relevant, please consider implementing the same as a plugin.

**Pull request closed without merging**